# Plotting selected GE of selected genes in heatmap for patients with RNASeq data
# nov22 2022
# Lilina Marie Boll
# lboll@imim.es


# the fields we want to plot are
# 1. immune infiltration
# 2. peptide presentation (MHC II, PD-L1)
# 3. IFN-gamma
# 4. TGF-beta
# 5. StromaEMT

# scale by row
# clonal TMB as annotation bar above / below heatmap

library(dplyr)
library(ggplot2)
library(RColorBrewer)
library(pheatmap)
library(readxl)


source('/genomics/users/lilian/Data/WES/generalfunctions.R')

datadir='/genomics/users/lilian/Data/RNA_Seq/'
outdir='/genomics/users/lilian/Data/RNA_Seq/02_Heatmap'
setwd(outdir)


# Load TMM CPM Robject generated by Jùlia
load ('/bicoh/joaquim/immuno_serie/cnag/RNASeq/03_COUNTS/Boll-Perera-Bel/TMM.logCPM.matrix.RData')
# colnames as Patient.ID
colnames(cpm.matrx) <- as.character(sub("_.*", "", colnames(cpm.matrx))) # 26 rows due to post samples
# remove post treatment RNA samples:
cpm.matrx  = subset(cpm.matrx, select = -c(NR3p,R9p,NR5p,R13p,R19p) )

# Sort by response: NR, PR, CR
cpm.matrx <- cpm.matrx[, c("NR1", "NR3"  ,"NR4"  ,"NR5"  ,
                         "NR6" ,"NR7"  ,"NR8" ,"R1" ,"R2"  ,"R3"  ,
                         "R4", "R5", "R8", "R10", "R12", "R13", 
                         "R17","R6"  , "R7", "R15" )]
PatientIDs <- colnames(cpm.matrx)
length(PatientIDs) # 20

# load mutation data for annotation
mutation_data <- read_xlsx('/genomics/users/lilian/Data/WES/Boll-Perera-Bel_datafile.xlsx',
          sheet='mutation_data', range="A1:Q28")


# ======================== DECIDE WHICH GENES TO PLOT  =======================

# Obtain list of genes to be included in heatmap:
# From Jùlia's research on gene sets:
Tcell_inflammed <- c("CD3D","ID01","CIITA","CD3E","CCL5","GZMK","CD2","HLA-DRA","CXCL13","IL2RG","NKG7","HLA-E","CXCR6","LAG3","TAGAP","CXCL10","STAT1","GZMB")

stromaEMT <- c("COL3A1","COL5A2","COL5A1","FBN1","COL1A1","FN1","COL6A3","SERPINE1","COL1A2","COL4A1","COL4A2","VCAN","IGFBP3","TGFBI","SPARC","LUM","LAMC1","LOX","LAMC2",
               "CCN2","TAGLN","COL7A1","LOXL2","COL6A2","ITGAV","THBS2","COL16A1","NNMT","TPM1","CDH2","MMP2","COL11A1","THBS1","FAP","BGN","SERPINH1","FSTL1","POSTN",
               "THY1","SPP1","TNC","TFPI2","NID2","ITGB5","MMP3","VIM","LOXL1","FBLN5","COL12A1","ELN","CDH11","COMP","SPOCK1","BMP1","IL32","LAMA3","TIMP1","QSOX1","TIMP3",
               "VCAM1","CCN1","EDIL3","CALD1","MAGEE1","FBLN1","SGCB","ECM1","LAMA2","FSTL3","TPM2","INHBA","DAB2","EMP3","BASP1","ITGA5","MGP","VEGFA","CXCL1","WNT5A",
               "SDC1","PLOD2","PCOLCE","GREM1","ITGB1","COL5A3","RHOB","HTRA1","FGF2","SNTB1","GADD45A","MEST","LRRC15","TNFRSF11B","CD59","ACTA2","EFEMP2","MATN2","PCOLCE2",
               "SERPINE2","GPC1","ABI3BP","FUCA1","SLIT3","LAMA1","PMEPA1","COL8A2","FBN2","IGFBP2","PFN2","SDC4","CD44","GADD45B","CXCL8","GLIPR1","ANPEP","P3H1","VEGFC"
               ,"MMP14","SGCD","PLOD1","MATN3","MYL9","SLC6A8","CALU","PRRX1","TNFRSF12A","FMOD","ID2","GEM","PLAUR","MYLK","TGFB1","SFRP1","PLOD3","IL6","APLP1","FBLN2","MSX1",
               "PTX3","FZD8","JUN","FERMT2","DKK1","SNAI2","DST","TPM4","DCN","GJA1","PMP22","IGFBP4","COPA","LRP1","ITGA2","FLNA","MFAP5","PTHLH","TGFBR3","SFRP4","LGALS1",
               "RGS4","CDH6","SAT1","NT5E","DPYSL3","PPIB","TGM2","SGCG","ITGB3","PDLIM4","CTHRC1","ECM2","CRLF1","AREG","IL15","MCM7","GAS1","PRSS2","CADM1","OXTR","SCG2",
               "CXCL6","MMP1","TNFAIP3","CAPG","CAP2","MXRA5","FOXC2","NTM","ENO2","FAS","BDNF","ADAM12","PVR","CXCL12","PDGFRB","SLIT2","NOTCH2","COLGALT1","GPX7","WIPF1")
# Mariathasan et al., 2018
TGFb.mariathasan <-c("TGFB1", "TGFBR2", "ACTA2", "COL4A1", "TAGLN", "SH3PXD2A")
IFNg.cristescu <- c("CCL5","CD27","CD276","CD8A","CMKLR1","CXCL9","CXCR6","IDO1","LAG3","NKG7","PSMB10","STAT1")
                # excluded: "PDCD1LG2", "TIGIT", "CD274" (in ICP.mariathasan)
                # excluded "HLA-DQA1","HLA-DRB1","HLA-E", (in MHCII)
IFNg.ayers <- c("ID01","CXCL10","CXCL9","STAT1","IFNG") # "HLA-DRA" because in GO-MHC

MHC.II <- c("CIITA", "CD74", "HLA-DPA1", "HLA-DPB1", "HLA-DPB2", "HLA-DQA1", "HLA-DRB1", "HLA-DRB5", "HLA-DRB6", "CTSH", "NCOA1")
GO.POS.REG.MHCII<- c("AZU1","CIITA","IFNG","IL10","IL33","IL4","JAK2","SIRT1","TLR4","TMEM106A","XBP1")
GO.MHCII<- c("CD74","HLA-DMA","HLA-DMB","HLA-DOA","HLA-DOB","HLA-DPA1","HLA-DPB1","HLA-DQA1","HLA-DQA2","HLA-DQB1","HLA-DQB2","HLA-DRA",
             "HLA-DRB1","HLA-DRB3","HLA-DRB4","HLA-DRB5")
GO.PRESENTING.MHCII<- c("ACTR10","ACTR1A","ACTR1B","AP1B1","AP1G1","AP1M1","AP1M2","AP1S1","AP1S2","AP1S3","AP2A1","AP2A2","AP2B1","AP2M1","AP2S1","ARF1","CANX","CAPZA1",
                        "CAPZA2","CAPZA3","CAPZB","CD74","CENPE","CLTA","CLTC","CTSD","CTSE","CTSF","CTSL","CTSS","CTSV","DCTN1","DCTN2","DCTN3","DCTN4","DCTN5","DCTN6",
                        "DNM2","DYNC1H1","DYNC1I1","DYNC1I2","DYNC1LI1","DYNC1LI2","DYNLL1","DYNLL2","FCER1G","FCGR2B","HLA-DMA","HLA-DMB","HLA-DOA","HLA-DOB","HLA-DPA1",
                        "HLA-DPB1","HLA-DQA1","HLA-DQA2","HLA-DQB1","HLA-DQB2","HLA-DRA","HLA-DRB1","HLA-DRB3","HLA-DRB4","HLA-DRB5","IFI30","KIF11","KIF15","KIF18A",
                        "KIF22","KIF23","KIF26A","KIF2A","KIF2B","KIF2C","KIF3A","KIF3B","KIF3C","KIF4A","KIF4B","KIF5A","KIFAP3","KLC1","KLC2","LGMN","MARCHF1",
                        "MARCHF8","OSBPL1A","PIKFYVE","PYCARD","RAB7A","RACGAP1","RILP","SAR1B","SEC13","SEC23A","SEC24A","SEC24B","SEC24C","SEC24D","SEC31A","SH3GL2",
                        "SPTBN2","THBS1","TRAF6","TREM2")
                # excluded "LAG3" (in IFNg.cristescu)


# from Mariathasan 2018 et al., Figure 3
TGfb.mariathasan<- c("TGFB1","TGFBR2", "ACTA2", "COL4A1", "TAGLN", "SH3PXD2A")
CD8Teff.mariathasan <- c("CD8A","GZMA","GZMB","PRF1","CXCL9","CXCL10","TBX21")
ICP.mariathasan <- c("CD274", "PDCD1LG2","CTLA4","PDCD1","HAVCR2","TIGIT") # excluded: LAG3 (in IFNg.cristescu)
EMT.mariathasan <- c("CLDN3","CLDN7","CLDN4","CDH1","VIM","TWIST1","ZEB1","ZEB2")
F_TBRS.mariathasan <- c("ACTA2","COL4A1","TAGLN","SH3PXD2A")

FGFR3.mariathasan <- c("FGFR3", "TP53", "WNT7B")
APM.mariathasan <- c("TAP1","TAP2","B2M","HLA-A","HLA-B","HLA-C")
cellCycle.mariathasan <- c("MKI67","CCNE1","BUB1","BUB1B","CCNB2","CDC25C","CDK2","MCM4","MCM6","MCM2")
histone.mariathasan <- c("HIST1H2AI","HIST1H2AG","HIST1H2BL","HIST1H2BF")
DDR.mariathasan <- c("BRCA2","ERCC2","ERCC4","FANCA","FANCB","FANCD2","PALB2","POLE","RAD51C")
Angio.mariathasan <- c("TEK","CDH5","SOX17","SOX18")

microRNAcancer.mariathasan <- c("BRCA1", "CCNE1", "CCNE2", "CDC25A", "CDC25C", "CDCA5", "CDKN2A", "DNMT1", "E2F1", "E2F2", "EZH2", "KIF23", "STMN1", "TRIM71")
RNAdegredation.mariathasan <- c("CNOT10", "EXOSC2", "EXOSC8", "LSM3", "LSM4", "LSM5", "PARN")

# Other important genes
immuneinflitr.boll <- c("CD45", "CXCL13")
ICP.boll <- c("CCND1" )

#merge ALL to final gene lists, remove duplicates
ICP <- ICP.mariathasan %>% append(ICP.boll) %>% unique()
ICP.MHC <- ICP %>% append(GO.PRESENTING.MHCII) %>% unique()
IFN.g <- IFNg.cristescu %>% append(IFNg.ayers) %>% unique()
TGfb <- TGfb.mariathasan %>% append(TGFb.mariathasan) %>% unique()
stromaEMT <- EMT.mariathasan %>% unique()
# ======================== BUILD Gene Expression MATRIX  =======================

# Get gene expression from TMM cpm Robject
get_geneExp <- function(pathway) {
  df <- data.frame(matrix(nrow = 0, ncol = length(PatientIDs))) 
  colnames(df) <- PatientIDs
  genelist <- c()
  for (gene in pathway){
    if (gene %in% rownames(cpm.matrx)) {
      # find row mathcing gene name and keep logFC and adj.P.Val
      df_tmp <-  cpm.matrx[rownames(cpm.matrx)==gene,]
#      df <- rbind(df, df_tmp)
      df[nrow(df)+1,] <- df_tmp
      genelist <- append(genelist, gene)
    } else{
      print(paste(gene,"was not found in differential expressed gene tabel!"))
    }
  }
  rownames(df) <- genelist
  return (df)
}


# FINAL SELECTION:
ICP.MHC <- sort(c("CD274","CD279","CTLA4","PDCD1", "PDCD1LG2", "CCND1","CD45", "CXCL19", "TIGIT")) # "HAVCR2",
IFN.g <- sort(c( "CD27",  "CXCR6", "IFNG","IFNG1","LAG3","STAT1", "ID01", "NKG7"  )) #"CD8A"  "CXCL9"  "CMKLR1" ,"PSMB10","CCL5","CD276",
immmuneinfiltr <- sort(c("CD8A","GZMA","GZMB","CXCL9","CXCL10")) # "GZMK",
# "TBX21", "PRF1","CD3D","CD3E","CIITA", "TAGAP" # CD8Teff + Tcell_inflammed
MHCII <- c("CIITA1", "B2M","CD74","HLA-DMA","HLA-DPA1","HLA-DQA2") # "HLA-DRA" "HLA-DQA1","HLA-DQB1","HLA-DOB","HLA-DMB","HLA-DQB2" "HLA-DPB1","HLA-DOA",

TGfb <- sort(c("CCL24","IL4R","JUNB","TGFB1","SKI","PPP1R15A")) 
# "PDGFA" ACVR1 "IFNGR1","KIT","TNFRSF1A","TNFRSF10B","IL6ST","TGFBR2", "TNFRSF14", "FLT4","PDGFRB","TAGLN" TGDBR2 "COL4A1","LIF","JUN","NCOR2"

stromaEMT <- sort(c("CALD1","COL6A2","ELN","GAS1","ITGA5")) 
# "MYL9","TPM2", "BDNF","COL16A1""RHOB","CXCL1","CLDN7", "VIM","TWIST1","ZEB1","ZEB2" )) # CLDN4, "CLDN3", "CDH1",,"TPM1""GADD45B",,"MYLK"

# significant genes from ORA hallmark EMT
# BDNF/CALD1/COL16A1/COL6A2/CXCL1/ELN/GADD45B/GAS1/ITGA5/JUN/MYL9/MYLK/RHOB/TPM1/TPM2

# check if any genes are duplicated in two pathways now:
merged_genelists <- c(ICP.MHC, IFN.g, MHCII, immmuneinfiltr, TGfb, stromaEMT)
merged_genelists[duplicated(merged_genelists)==TRUE]
# should be 0


GE_ICP <- get_geneExp(ICP.MHC)
GE_IFNg <- get_geneExp(IFN.g)
GE_immmuneinfiltr <- get_geneExp(immmuneinfiltr)
GE_MHCII <- get_geneExp(MHCII)
GE_TGfb <- get_geneExp(TGfb)
GE_stromaEMT <- get_geneExp(stromaEMT)

GE_selected <- do.call("rbind", list(GE_ICP, GE_IFNg, GE_immmuneinfiltr, GE_MHCII, GE_TGfb, GE_stromaEMT))
matrix.GE_selected <- as.matrix(GE_selected)
# adding row with pathways names
GE_ICP$pathway <- "Immune checkpoints"
GE_IFNg$pathway <- "IFNg"
GE_immmuneinfiltr$pathway <- "Immune infiltrated"
GE_MHCII$pathway <- "MHC II"
GE_TGfb$pathway <- "TGFb"
GE_stromaEMT$pathway <- "EMT"

GE_selected <- do.call("rbind", list(GE_ICP, GE_IFNg,GE_immmuneinfiltr, GE_MHCII, GE_TGfb, GE_stromaEMT))

# sort by Response (NR, PR, CR)
GE_selected <- GE_selected[, c("NR1", "NR3"  ,"NR4"  ,"NR5"  ,"NR6" ,"NR7"  ,"NR8" ,"R1" ,"R2"  ,"R3"  , "R4", 
                               "R5", "R8", "R10", "R12", "R13"  , "R17","R6"  , "R7", "R15", "pathway" )]

# ============================= PLOT HEATMAP  ==================================

# --------------------------- pheatmap ---------------------

table(GE_selected$pathway)
# cluster by pathway
pathway_col <- data.frame(Pathway = GE_selected$pathway, row.names = rownames(GE_selected))

# add response
response_col <- data.frame(Response = c(rep("NR", 7), rep("PR", 10), rep("CR", 3)))
row.names(response_col) <- colnames(GE_selected)[c(1:20)]

TMB_col <- data.frame(mutation_data[,c("TMB.clonal","TMB.total")])
rownames(TMB_col) <- mutation_data$Patient.ID

# merge TMB and Response in sample_col
sample_col <- merge(response_col, TMB_col, by=0, all=F)

row.names(sample_col) <- sample_col$Row.names

sample_col <- sample_col[,c("Response","TMB.clonal", "TMB.total")]
sample_col$TMB.total <- as.integer(sample_col$TMB.total)
sample_col$TMB.clonal <- as.integer(sample_col$TMB.clonal)
colnames(sample_col) <- c("Response", "clonal TMB","TMB")

# == COLOURS ==
my_colour = list(
  Response = c(NR="#92C5DE",PR= "#D17373", CR="#963c3c" ),
  Pathway = c('Immune checkpoints' = "#FDDDA0", IFNg = "#9A8822", 'Immune infiltrated'="#a9d6db",  'MHC II'="#b35050" , TGFb='#74A089', EMT = "#F8AFA8"),
  'clonal TMB'=colorRampPalette(c("#ebfffd","#4b9992"))(20),
  TMB=colorRampPalette(c("#ebeef7","#3b41b8"))(20)
)

# reorder the matrix based on response
matrix.GE_selected_ordered <- matrix.GE_selected[, c("NR1", "NR3"  ,"NR4"  ,"NR5"  ,"NR6" ,"NR7"  ,"NR8" ,"R1" ,"R2"  ,"R3"  ,
                                                     "R4", "R5", "R8", "R10", "R12", "R13"  ,"R17","R6"  , "R7", "R15" )]

pheat <- pheatmap(matrix.GE_selected_ordered, 
                  scale="row",
                  # annotations
                  annotation_row = pathway_col,
                  annotation_col = sample_col, 
                  annotation_colors = my_colour,
                  # keep order of cols and rows
                  cluster_cols=F, cluster_rows = F,
                  # aestetics
                  gaps_row = 21,  gaps_col = 7,
                  border_color = "grey60")

save_pheatmap_pdf <- function(x, filename, width=8, height=7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

save_pheatmap_pdf(pheat, "heatmap.pdf")

# relate TMB / clonal TMB with CD8A or CXCL9 or PDL1 / CTL4A
cpm.matrx.4 <-cpm.matrx[rownames(cpm.matrx)=='CD8A',]
cpm.matrx.4 <- rbind(cpm.matrx.4, cpm.matrx[rownames(cpm.matrx)=='CXCL9',])
cpm.matrx.4 <- rbind(cpm.matrx.4, cpm.matrx[rownames(cpm.matrx)=='PDCD1',])
cpm.matrx.4 <- rbind(cpm.matrx.4, cpm.matrx[rownames(cpm.matrx)=='CTLA4',])
dim(cpm.matrx.4)
rownames(cpm.matrx.4) <- c('CD8A', 'CXCL9', 'PDCD1', 'CTLA4')
cpm.matrx.4

df.cpm.tmm <- as.data.frame(t(cpm.matrx.4))

mut_cpm <- merge(x=df.cpm.tmm, y=mutation_data, by.x=0, by.y="Patient.ID", all=F)
dim(mut_cpm)
colnames(mut_cpm)
names(mut_cpm)[1] <- "Patient.ID"
mut_cpm <- add_response(mut_cpm)



#CD8A.TMB = 
ggplot(mut_cpm,aes(TMB.clonal,CXCL9,  col = Response, label=Patient.ID)) +
  geom_text( size = 4) +
  geom_point(size=5)+
  # geom_rug()+
  scale_color_manual(values = resp_colour, labels=c("Responders", "Non-responders")) +
  theme_minimal(base_size = 14) +
  labs(x = "CXCL9 gene expression", y = "clonal TMB") +
  theme(legend.title = element_blank(), legend.position = "bottom") 



library(ggrepel)
  geom_label_repel(aes(label = Patient.ID),
                   box.padding   = 0.35, 
                   point.padding = 0.5,
                   segment.color = 'grey50')




ggsave(p, "test.png")


  theme_minimal(base_size = 14) +
  theme(plot.title = element_text(size=16)) +
  ggtitle("TMB by ICI response") +
  xlab("TMB") + ylab("CD8A gene expression")+
  scale_fill_manual(values=resp_colour)

  
  